### PR TITLE
Loop promotion analysis WAR/shortcut and cycle check

### PR DESCRIPTION
**Pull request by naoyam**
To workaround the cyclic graph issue (#3072) as much as possible, this PR skips the full analysis when possible. When skipped, the alternative analysis is just simply picking the first ID, so it shouldn't have any constraint with the graph structure. It can be also considered a shortcut as it would skip the analysis if not necessary. 

Also added a check of existence of cycles in a graph as it isn't supported.

### CODE REVIEW

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
!build

File: csrc/device_lower/utils.h [hunk 1]
```diff
@@ -16,7 +16,6 @@
 #include <kernel_ir.h>
 #include <parallel_type_bitmap.h>
 #include <val_graph.h>
-#include <val_graph_visitor.h>
```
Comment (naoyam) [inline csrc/device_lower/utils.h:19]: Cleaned up the usage of this header file as I noticed just changing this file required recompilation of more than 160 files..

**Comment by naoyam** [comment]
Hmm, I didn't expect this would cause any error. Making this draft for now.

**Comment by naoyam** [comment]
!build

File: csrc/id_model/loop_promotion.cpp [hunk 2]
```diff
@@ -898,4 +932,60 @@ std::unordered_map<ValGroup, IterDomain*> LoopPromotionMapBuilder::get(
...
+      if ((is_const && !this_is_const) ||
+          (is_const == this_is_const && this_num_exprs >= num_exprs)) {
+        continue;
+      }
+
+      // Update the current promotion
+      is_const = this_is_const;
+      promotion = loop_id;
+      num_exprs = this_num_exprs;
+    }
```
Comment (naoyam) [inline csrc/id_model/loop_promotion.cpp:981]: This part is completely optional. Just make it easier to my eyes when looking at debug print and also it avoids this test error. https://github.com/NVIDIA/Fuser/pull/3079#issuecomment-2387571701

**Review by zasdfgbnm — APPROVED** [review]

**Comment by naoyam** [comment]
!build

**Pull request merged**